index.html

```diff
@@ -0,0 +1,201 @@
+
+## Spec URLs
+
+```graphql
+#::[Declaration of using_URL scalar]
+scalar using_URL
+  @specifiedBy(url: "https://apollographql.com/docs/specs/using#spec-urls")
+```
+
+The URL of a specification.
+
+Specification URLs **must** be valid [RFC 3986 URLs](https://tools.ietf.org/html/rfc3986). They **should** provide the content of the appropriate version of the specification in some human-readable form. It is not required that the content be machine-readable in any particular way.
+
+The final component of the URL's `"/"`-separated `path` **must** be a valid [SemVer specifier](#semver-specifiers) identifying the version of the spec in use. Everything prior to the version specifier is the spec's *base URL*. Processors **should** treat the spec's base URL as an opaque string identifying the specification.
+
+This spec assigns no meaning to additional URL components, such as the query string or fragment. Processors **should** ignore any meaningless URL components when parsing the base URL and semver specifier.
+
+```graphql
+#::[`@using` directive with a query string]
+schema
+  @using(spec: "https://spec.example.com/exampleSpec/1?unrelated=value")
+  { ... }
+```
+
+This directive:
+- Declares that the document conforms to a spec which can be found at `https://spec.example.com/exampleSpec/1`
+- Has a base URL of `https://spec.example.com/exampleSpec`
+- Has a [semver specifier](#semver-specifier) of `1`
+- Has a query string of `"unrelated=value":
+  - This has no effect unless `exampleSpec` assigns it specific meaning.
+
+## Semantic Versioning
+
+Specifications **should** provide a semantic version number.
+
+### SemVer Specifiers
+
+The final `/`-separated path component **must** be a semver specifier. Semver specifiers take the form `v*major*.*minor*.*patch*-*prerelease*`. The minor, patch, and pre-release specifiers are all optional. Examples:
+
+```
+#::[Valid SemVer specifiers]
+v1
+v1.2
+v1.2.3-alpha
+v2.2.3-alpha.2
+```
+
+### SemVer selection
+
+Processors **should** select the highest available spec implementation which is compatible with the requested version according to the [SemVer 2.0.0](https://semver.org/spec/v2.0.0.html) rules for version compatibility. For example, consider this directive:
+
+```graphql
+#::[`@using` version 2.x.y]
+schema @using(spec: "https://specs.example.com/exampleSpec/v2")
+```
+
+Upon seeing this directive, processors which support the `https://specs.example.com/exampleSpec` spec **should** activate a version of their implementation which is compatible with major version 2.
+
+```graphql
+#::[`@using` version 3.1.1-alpha]
+schema @using(spec: "https://specs.example.com/exampleSpec/v3.1.1-alpha")
+```
+
+This selects versions in the `3.1.1-alpha` prerelease series. For example, a processor may satisfy this requirement by activating version `3.1.1-alpha.2` of `exampleSpec`.
```

Comment:
I don't think Semver v2.0.0 makes any claims about "compatibility" requirements for pre-release versions.

> A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version.

It provides an order for them, but that's not the same as compatibility.

I think it's fine if we want to define this compatibility operation but we have to actually explicitly define it rather than handwave at Semver.

I would be tempted to say that it would be better to just say that a schema needs to name the specific version of the spec (ie a full semver) that it's designed for, and leave the logic of "given a claim that I'm written for v1.2.3, how should I process it" up to the consumer rather than bringing the concept of of compatibility ranges into the actual `@using` spec. I'm honestly confused by the concept of claiming that a concrete document follows a prefix range of potential specs (some of which may not have been released yet) and I don't understand the advantage over just saying "my software produces docs that fit this precise version of the spec and it's your job to know which implementation wants that version". But maybe I'm missing something! I do think mandating that the URL contain a version is a good idea I would not have thought of.

Reply:
I agree with David here in that it's probably better to leave the compatibility question of how to process a GraphQL document written using a specific version of a spec to document consumers and not the document producer.

It could be useful if the producer understands their consumers relatively well and wants to impose restrictions because the consumers aren't doing what the producer wants (e.g. they're too lax, or they make assumptions because the spec being used has some issues/is vague), although in practice I'd imagine producers don't want to look into that, and that consumers would hopefully handle complexities in the spec properly (weighing the tradeoffs, and forwarding decisions to users where needed). Similarly, it's useful in that it forces consumers to handle version compatibility instead of doing something naive like just matching the version string against a list, but we could also just require that consumers handle SemVer version compatibility in a particular way.

So basically we'd change the `@using` spec to just reference a single SemVer version, but we'd add a section to the spec to talk about how consumers should handle version compatibility. This would effectively leave it up to document consumers to deal with the complexities of spec versioning and not document producers. This is probably better since consumers know more about the application and have the right information to make compatibility decisions, while producers usually don't. There's also a factor here where producers may often be humans while consumers are less often humans (usually they're libraries written by humans), so less humans overall have to think about version compatibility if we place the burden on consumers.

Reply:
Was a decision made to leave this alone? I don't really see the value of having a custom semver range algorithm baked into the spec rather than letting schemas brand themselves with the specific version they implement and letting consumers do their own ranging.

Reply:
This section hasn't been updated yet. Language in the earlier sections has been updated to specify that the version in the url should be the exact version of the spec against which the document was authored.

I think we want to be one step more involved than just punting to the consumers. Consider the case of a composer operating on two input schemas, which both reference some unknown spec `a`:

```
schema @using(spec: "https://some.spec/a/v1.0")
```

and

```
schema @using(spec: "https://some.spec/a/v1.3")
```

I think the composer should be able to merge these `@using` directives into:

```
schema @using(spec: "https://some.spec/a/v1.3")
```

But to get that behavior, we need to specify at least the major/minor facets of semver (i don't think patch versions are ever useful in documents, but feel free to change my mind there).

Reply:
(Will think about this next week)